dwarf-fortress: update to 50

numinit opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

`This is a pull request. All craftsdwarfship is of the highest quality. It menaces with spikes of Nix and bash. On the pull request are several derivations.`

- [x] Support Dwarf Fortress 50
  - [x] Look at the DF data directory and see if our symlinking approach still works.
  - [x] Fix audio. The library loads but I get nothing...
  - [x] REALLY fix audio [tbd: is there a fix?]
  - [x] Fix dfhack. It compiles but doesn't run.
  - [x] Fix dfhack upstream too, maybe. Their build system tries to download from the internet now and I had to hack around it. (upstream PR: https://github.com/DFHack/dfhack/pull/4418)
  - [x] Add more DF 50 versions.
  - [x] installCheckPhase. This is an extremely complex set of packages!

- [x] Fix Dwarf Fortress Unfuck for DF 0.47 (upstream ticket: https://github.com/svenstaro/dwarf_fortress_unfuck/issues/29). This is not required for DF 50 and above. Neither is TWBT, though we still support it for previous versions.

### How to test

`NIXPKGS_ALLOW_UNFREE=1 nix build .#dwarf-fortress-packages.dwarf-fortress-full --impure -L --show-trace`
`NIXPKGS_ALLOW_UNFREE=1 nix-build -E '(import ./. {}).dwarf-fortress-packages.dwarf-fortress-full.override { dfVersion = "0.47.05"; }' --impure`

Maybe a couple other versions.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
dwarf-fortress: update to 50

numinit commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -148,29 +155,69 @@ in
+      ++ lib.optional (!isV50) SDL
+      ++ lib.optionals enableStoneSense [ allegro5 libGLU libGL ];
+
+    preConfigure = lib.optionalString isV50 ''
```
These shouldn't have the lib.* prefix

numinit commented on pkgs/games/dwarf-fortress/dwarf-therapist/default.nix:
```diff
@@ -6,17 +6,23 @@
-  version = "41.2.2";
+
+  # 41.2.5 is the last version to support Dwarf Fortress 0.47.
+  version = if isV50 then "42.1.5" else "41.2.5";
```
Not a typo.

SuperSandro2000 replied:
This seems really hacky. Can't we use an override?

jonringer commented on pkgs/games/dwarf-fortress/wrapper/dwarf-fortress-init.in:
```diff
@@ -30,16 +38,19 @@ forcecopy_path() {
 cd "$env_dir"
-for i in data/init/* data/!(init|index|announcement) raw; do
-  update_path "$i"
+for path in libfmod_plugin.so libsdl_mixer_plugin.so data/init/* data/!(init|index|announcement) raw libs dwarfort; do
+  if [ -e "$path" ]; then
+    update_path "$path"
+  else
+    cleanup_path "$path"
+  fi
```
```
error: builder for '/nix/store/1hzq17s5wqdr843iwbpn861ndfvq48h9-dwarf-fortress-config.drv' failed with exit code 1;
       last 1 log lines:
       > cp: cannot stat '/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/init.txt': No such file or directory
       For full logs, run 'nix log /nix/store/1hzq17s5wqdr843iwbpn861ndfvq48h9-dwarf-fortress-config.drv'.
error: 1 dependencies of derivation '/nix/store/nv9zlhlk5n30mnbrpaysy4jymj3y6p3k-dwarf-fortress-env-50.12.drv' failed to build
error: 1 dependencies of derivation '/nix/store/4hi7q2aicspyhyzm7pigd8yypgfzw8jb-dwarf-fortress-init.drv' failed to build
error: 1 dependencies of derivation '/nix/store/vx7mzqh435j9sfk583l45xkm7ybfr82w-dwarf-fortress-50.12.drv' failed to build
```

jonringer replied:
```
$ ls /nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/*
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/announcements.txt
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/arena.txt
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/colors.txt
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/d_init_default.txt
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/init_default.txt
/nix/store/np8f97rfbgwg465kpk2277r3azshig18-dwarf-fortress-base-env-50.12/data/init/interface.txt
```

numinit replied:
What derivation are you building? I think I see what the issue is:

```
  '' + forEach settings' (file: kv: ''              
    file=data/init/${lib.escapeShellArg file}.txt
    if [ -f "$file" ]; then # added
      cp ${baseEnv}/"$file" "$out/$file"
    fi     # added
```

But I'm curious where this is from (it's not the runtime of the wrapper, it's the build time...)

numinit replied:
Give it a shot now, I think I reproduced a case similar to yours and was able to fix it.

ncfavier replied:
Still getting this with just `nix-build -A dwarf-fortress`.

jonringer replied:
The default:
```
Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/300402

1 package failed to build:
dwarf-fortress
```

numinit replied:
Oooh, I see what's up. That derivation is trying to build a bunch of different versions.

numinit replied:
Give it a shot now.

numinit commented:
I've marked this as ready for review. It adds new functionality and could be used as a base to clean up this package a bit. The three-wrapper-deep style of this set of packages was originally very complex (but also exists for a reason, since we need to monkeypatch the expected binary MD5s in Therapist and dfhack, as well as create a game runtime directory that is symlinked into the store with directories for user data). Suggestions and further work on improving the style are welcome.

numinit committed:
dwarf-fortress: update release notes

ncfavier commented:
Still no sound for me in either `dwarf-fortress` or `dwarf-fortress-full`.

numinit commented:
Hrm, I'll look into it. I'm noticing some other jank but strace and ltrace couldn't pin it down.

numinit committed:
dwarf-fortress: df_permission: Renew permission from Tarn for v50

Get permission from Tarn to publish a v50 Dwarf Fortress in nixpkgs.
The last time we did this was in 2010. (moved to df_permission/2010)

numinit commented:
@ncfavier

The audio problem was unfortunately extremely complicated, despite just having to do with fmod being broken. Since Dwarf Fortress loads the fmod plugin if it can find it, and we link it into the data directory, we needed a new system to pass nixpkgs-specific options to the Dwarf Fortress wrapper so we don't link fmod if we don't want it (now the default; we can set `NIXPKGS_DF_OPTS=fmod` if we really want it). Of course, there are tests for this behavior now.

Along the way, I discovered a bunch of ambient paths in the dfhack wrapper, so had to fix that too, with another wrapper. Also unfortunately, the dfhack initscript assumes that the script directory is the one we want to be in, so I had to make a .dfhackrc in the dfhack build that cd's back into `NIXPKGS_DF_HOME` so things actually load. One more step to the Rube Goldberg; luckily dfhack appears to allow a distro packaging specific dfhackrc (basically) in addition to looking at ~/.dfhackrc.

Finally, I made sure we had permission from Tarn Adams to package v50, since the last time we asked was 14 years ago, and v50 was also released on Steam. He was fine with it continuing to be packaged; the email is in the df_permission folder along with the previous one.

The upshot is that audio with SDL2_mixer works now. Oy.

SuperSandro2000 commented on pkgs/games/dwarf-fortress/default.nix:
```diff
@@ -46,31 +46,28 @@ let
 
```
Please remove the v prefix from this commit

SuperSandro2000 commented on pkgs/games/dwarf-fortress/default.nix:
```diff
@@ -46,31 +46,28 @@ let
   # out, ensure that (unfuck|dfhack|twbt) are all up to date before changing
-  # this.
-  latestVersion = "0.47.05";
+  # this. Note that unfuck and twbt are not required for v50.
```
```suggestion
  # this. Note that unfuck and twbt are not required for 50.
```

SuperSandro2000 commented on pkgs/games/dwarf-fortress/default.nix:
```diff
@@ -46,31 +46,28 @@ let
       value =
         let
-          # I can't believe this syntax works. Spikes of Nix code indeed...
+          isV50 = lib.versionAtLeast dfVersion "50.0";
```
```suggestion
          isAtLeast50 = lib.versionAtLeast dfVersion "50.0";
```

SuperSandro2000 commented on pkgs/games/dwarf-fortress/default.nix:
```diff
@@ -46,31 +46,28 @@ let
-          # I can't believe this syntax works. Spikes of Nix code indeed...
+          isV50 = lib.versionAtLeast dfVersion "50.0";
+
+          dwarf-fortress-unfuck = if isV50 then null else callPackage ./unfuck.nix { inherit dfVersion; };
```
We should work with lib.optionalAttrs instead of setting this to null.

SuperSandro2000 commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -80,6 +86,7 @@ let
     else throw "[DFHack] Unsupported Dwarf Fortress version: ${dfVersion}";
 
   version = release.dfHackRelease;
+  isV50 = lib.versionAtLeast version "50.0";
```
same here

SuperSandro2000 commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -148,29 +155,69 @@ in
       sed -i 's@cached_path = path_string.*@cached_path = getenv("DF_DIR");@' library/Process-linux.cpp
+
+      # shadows something in cmath.h
+      sed -i 's@lerp@_lerp@g' plugins/remotefortressreader/remotefortressreader.cpp
```
We should use substituteInPlace --replace-fail if possible to notice silent breakages

SuperSandro2000 commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -148,29 +155,69 @@ in
+    preConfigure = lib.optionalString isV50 ''
+      # Trick the build system into not downloading any files.
+      if [ -f CMake/DownloadFile.cmake ]; then
+        cat <<-EOF > CMake/DownloadFile.cmake
```
Should we maybe do a patch instead to make this easier to manage?

SuperSandro2000 commented on pkgs/games/dwarf-fortress/game.nix:
```diff
@@ -41,9 +44,21 @@ let
+  libpath = makeLibraryPath (
+    [ stdenv.cc.cc stdenv.cc.libc ]
+    ++ lib.optionals (!isV50) [ SDL ]
+    ++ lib.optionals enableUnfuck [ dwarf-fortress-unfuck ]
```
We could just use lib.optional here

SuperSandro2000 commented on pkgs/games/dwarf-fortress/game.nix:
```diff
@@ -57,25 +72,42 @@ let
 
+  sourceRoot = ".";
+
+  postUnpack = optionalString stdenv.isLinux ''
+    if [ -d df_linux ]; then
+      mv df_linux/* .
+    fi
+  '' + optionalString stdenv.isDarwin ''
+    if [ -d df_osx ]; then
+      mv df_osx/* .
+    fi
+  '';
```
Maybe it is easier to make the directory conditional?

SuperSandro2000 commented on pkgs/games/dwarf-fortress/game.nix:
```diff
@@ -57,25 +72,42 @@ let
     mkdir -p $out
     cp -r * $out
-    rm $out/libs/lib*
+    ${lib.optionalString (!isV50) "rm -f $out/libs/*.so"}
```
```suggestion
  '' + lib.optionalString (!isV50) ''
    rm -f $out/libs/*.so
  '' + ''
```
we can save ourselves string interpolation strugle

SuperSandro2000 commented on pkgs/games/dwarf-fortress/wrapper/default.nix:
```diff
@@ -31,10 +31,12 @@
     inherit enableStoneSense;
   };
 
+  enableTWBT' = enableTWBT && (twbt != null);
```
Why should twbt be null?

numinit replied:
TWBT doesn't even compile for version 50, and it's not needed anyway, so we shouldn't put it in the wrapper. I've improved the condition to check for the dfVersion as a passthru.

SuperSandro2000 commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -29,8 +29,8 @@ let
     optional
```
Please rebase this into the original commit to keep the history cleaner

SuperSandro2000 commented on pkgs/games/dwarf-fortress/dfhack/default.nix:
```diff
@@ -141,9 +141,13 @@ in
       url = "https://github.com/DFHack/dfhack/commit/7bdf958518d2892ee89a7173224a069c4a2190d8.patch";
```
also this commit can be rebased to make the history cleaner

SuperSandro2000 commented on pkgs/games/dwarf-fortress/game.nix:
```diff
@@ -135,8 +144,18 @@ stdenv.mkDerivation {
+
+  fixupPhase = ''
+    runHook preFixup
+    runHook postFixup
```
postFixup should be at the end

numinit replied:
Hmm... fixed it with an exit trap. It actually needs to run after the last fixup due to autoPatchelf.

SuperSandro2000 commented on pkgs/games/dwarf-fortress/wrapper/default.nix:
```diff
@@ -185,7 +192,35 @@ stdenv.mkDerivation {
+    expect ${dfHackExpectScript}
+  '' + ''
+    expect ${vanillaExpectScript}
+  '' + ''
```
```suggestion
    expect ${if enableDFHack then dfHackExpectScript else vanillaExpectScript}
```

SuperSandro2000 commented on pkgs/games/dwarf-fortress/df_permission/2010:
This file should get a patch file ending

ncfavier replied:
It's not a patch.

SuperSandro2000 commented on pkgs/games/dwarf-fortress/df_permission/2024:
```diff
@@ -0,0 +1,57 @@
+From: Tarn Adams <tarn.adams@gmail.com>
```
same here

numinit replied:
same, this is email permission from the developer to allow us to package it in nixpkgs

SuperSandro2000 commented:
GitHub ate my review comment:
Please cleanup the commit history. Many changes do get further fixups further down the line which makes following the history afterwards unnecessarily hard and also makes reviewing this PR hard.

ncfavier commented:
I've got sound in the menu now, but not in game (the "track playing" still says "classic DF title theme" but nothing plays).

Also, exiting the game results in

```
double free or corruption (out)
Aborted (core dumped)
```

(with the `pkgs.dwarf-fortress` derivation)

numinit committed:
dwarf-fortress: address code review feedback

numinit commented on pkgs/games/dwarf-fortress/game.nix:
```diff
@@ -118,15 +115,12 @@ stdenv.mkDerivation {
-    patchelf \
-      --set-interpreter $(cat ${stdenv.cc}/nix-support/dynamic-linker) \
-      --set-rpath "${libpath}" \
-      $exe
```
Not needed anymore, autoPatchelf works.

numinit commented:
@ncfavier re: audio

I compared my Steam directory vs. the vanilla DF data directory, and there's barely any sound in vanilla compared to the Steam version. It's the same way in 50.11 too. Strange about song_game.ogg not playing though...it plays on prior versions like 0.47.05.

I assume your menu looks like this?

![Screenshot_20240407_001626](https://github.com/NixOS/nixpkgs/assets/369111/8b250fba-fe7c-47c6-97f6-6691d0c3f7e3)

If I export `NIXPKGS_DF_OPTS=fmod` then the menu finally says:

![Screenshot_20240407_001540](https://github.com/NixOS/nixpkgs/assets/369111/87bccae4-8b04-41b8-bb69-13871b464b02)

Yet there's no audio at all... a couple friends with some non-NixOS linux setups will try it and see if it's broken without the packaging. Or worst case, I can do a qemu vm or something.

re: `double free or corruption (out)` I haven't reproduced this, although I _have_ seen division by zero errors when using dfhack on exit; it was enough to force an `exit 0` from the `expect` scripts after it prints what we expect, because exiting can be slightly broken in general...

numinit commented:
Ran all the wrapper scripts through a pretty strict shellcheck pass.

numinit committed:
dwarf-fortress: dfhack: update to 50.12-r3

numinit commented:
Update to release build of dfhack.

jonringer reviewed (approved):
Was able to launch DF and get audio.

Thanks for all the hard work, this looks like it took quite of bit of effort to update the update scripts and adjacent packages. <3

jonringer commented:
cc @SuperSandro2000 for another review. Would like to respect previous reviews
